### ISSUE TITLE
[boost] build failure with mingw and spaces in path

### ISSUE THREAD

**Issue by autoantwort** [microsoft/vcpkg#16218]
**Host Environment**
- OS: Windows
- Compiler: g++.exe (x86_64-posix-seh-rev0, Built by MinGW-W64 project) 8.1.0

**To Reproduce**
1. `cd C:/`
2. `mkdir "foo bar"
3. `cd "foo bar"`
4. `git clone ... + bootstrap`
5. `./vcpkg.exe install boost-system --triplet=x64-mingw-dynamic`

**Failure logs** 
```
-- Downloading https://github.com/boostorg/system/archive/boost-1.75.0.tar.gz -> boostorg-system-boost-1.75.0.tar.gz...
-- Extracting source C:/foo bar/vcpkg/downloads/boostorg-system-boost-1.75.0.tar.gz
-- Using source at C:/foo bar/vcpkg/buildtrees/boost-system/src/ost-1.75.0-2027312ef4.clean
-- Building release...
-- Configuring x64-mingw-dynamic
-- Building x64-mingw-dynamic-rel
-- Building debug...
-- Configuring x64-mingw-dynamic
-- Building x64-mingw-dynamic-dbg
CMake Error at installed/x64-mingw-dynamic/share/boost-build/boost-modular-build.cmake:159 (message):
  No libraries were produced.  This indicates a failure while building the
  boost library.
Call Stack (most recent call first):
  ports/boost-system/portfile.cmake:14 (boost_modular_build)
  scripts/ports.cmake:131 (include)


Error: Building package boost-system:x64-mingw-dynamic failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: boost-system:x64-mingw-dynamic
  Vcpkg version: 2021-01-13-768d8f95c9e752603d2c5901c7a7c7fbdb08af35

Additionally, attach any relevant sections from the log files above.
```

<details><summary><i>C:\foo bar\vcpkg\buildtrees\boost-system\install-x64-mingw-dynamic-dbg-out.log</i></summary>
<pre>[1/2] cmd.exe /C "cd /D "C:\foo bar\vcpkg\buildtrees\boost-system\src\ost-1.75.0-2027312ef4.clean\build" && "C:\foo bar\vcpkg\installed\x64-mingw-dynamic\tools\boost-build\b2.exe" "--user-config=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-dbg/user-config.jam" "--stagedir=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-dbg/stage" "--build-dir=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-dbg" runtime-link=shared link=shared address-model=64 architecture=x86 target-os=windows toolset=gcc threadapi=win32 -sZLIB_BINARY=z -sZLIB_LIBPATH="C:/Perl/c/lib/libz.a" -sBZIP2_BINARY=bz2d -sBZIP2_LIBPATH="BZIP2_LIBPATH-NOTFOUND" variant=debug --layout=system --with-atomic --with-random --with-date_time --with-filesystem --with-system --with-thread --with-chrono -sZLIB_INCLUDE="C:/Perl/c/include" -sBZIP2_INCLUDE="C:/Perl/c/include" -sICU_PATH="ICU_PATH-NOTFOUND" -j8 "-sBOOST_ROOT=C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build" "-sBOOST_BUILD_PATH=C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build" --debug-configuration --ignore-site-config --hash -q threading=multi debug-symbols=on stage"
notice: found boost-build.jam at C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build/boost-build.jam
notice: loading B2 from C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build/src/kernel/bootstrap.jam
notice: Site configuration files will be ignored due to the
notice: --ignore-site-config command-line option.
notice: Loading explicitly specified user configuration file:
    C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\user-config.jam
notice: Searching 'C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg' for user-config configuration file 'user-config.jam'.
notice: Loading user-config configuration file 'user-config.jam' from 'C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-dbg'.
notice: will use 'C:/Qt/Tools/mingw810_64/bin/x86_64-w64-mingw32-g++.exe' for gcc, condition <toolset>gcc-5.4.1
notice: using gcc libraries :: <toolset>gcc-5.4.1 :: C:/Qt/Tools/mingw810_64/bin C:/Qt/Tools/mingw810_64/lib C:/Qt/Tools/mingw810_64/lib32 C:/Qt/Tools/mingw810_64/lib64
notice: using gcc archiver :: <toolset>gcc-5.4.1 :: C:/Qt/Tools/mingw810_64/bin/ar.exe
notice: using gcc ranlib :: <toolset>gcc-5.4.1 :: C:/Qt/Tools/mingw810_64/bin/ranlib.exe
notice: using rc compiler :: <toolset>gcc-5.4.1 :: C:\Qt\Tools\mingw810_64\bin\windres.exe
...found 119 targets...
...updating 10 targets...
gcc.compile.c++ C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\boost\build\3a3c813ed341a12d5246d45ac0fd5eb0\error_code.o
gcc.link.dll.mingw C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\boost\build\3a3c813ed341a12d5246d45ac0fd5eb0\libboost_system.dll.a
common.copy C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\stage\lib\libboost_system.dll.a
C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\boost\build\3a3c813ed341a12d5246d45ac0fd5eb0\libboost_system.dll.a
        1 file(s) copied.
common.copy C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\stage\lib\libboost_system.dll
C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\boost\build\3a3c813ed341a12d5246d45ac0fd5eb0\libboost_system.dll
        1 file(s) copied.
...updated 10 targets...
[1/2] cmd.exe /C "cd /D "C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg" && "C:\foo bar\vcpkg\downloads\tools\cmake-3.19.2-windows\cmake-3.19.2-win32-x86\bin\cmake.exe" -P cmake_install.cmake"
-- Install configuration: "Debug"

</pre></details>

<details><summary><i>C:\foo bar\vcpkg\buildtrees\boost-system\install-x64-mingw-dynamic-rel-out.log</i></summary>
<pre>[1/2] cmd.exe /C "cd /D "C:\foo bar\vcpkg\buildtrees\boost-system\src\ost-1.75.0-2027312ef4.clean\build" && "C:\foo bar\vcpkg\installed\x64-mingw-dynamic\tools\boost-build\b2.exe" "--user-config=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel/user-config.jam" "--stagedir=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel/stage" "--build-dir=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel" runtime-link=shared link=shared address-model=64 architecture=x86 target-os=windows toolset=gcc threadapi=win32 -sZLIB_BINARY=z -sZLIB_LIBPATH="C:/Perl/c/lib/libz.a" -sBZIP2_BINARY=bz2 -sBZIP2_LIBPATH="C:/Perl/c/lib/libbz2.a" variant=release --layout=system --with-atomic --with-random --with-date_time --with-filesystem --with-system --with-thread --with-chrono -sZLIB_INCLUDE="C:/Perl/c/include" -sBZIP2_INCLUDE="C:/Perl/c/include" -sICU_PATH="ICU_PATH-NOTFOUND" -j8 "-sBOOST_ROOT=C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build" "-sBOOST_BUILD_PATH=C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build" --debug-configuration --ignore-site-config --hash -q threading=multi debug-symbols=on stage"
notice: found boost-build.jam at C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build/boost-build.jam
notice: loading B2 from C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build/src/kernel/bootstrap.jam
notice: Site configuration files will be ignored due to the
notice: --ignore-site-config command-line option.
notice: Loading explicitly specified user configuration file:
    C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\user-config.jam
notice: Searching 'C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel' for user-config configuration file 'user-config.jam'.
notice: Loading user-config configuration file 'user-config.jam' from 'C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel'.
notice: will use 'C:/Qt/Tools/mingw810_64/bin/x86_64-w64-mingw32-g++.exe' for gcc, condition <toolset>gcc-5.4.1
notice: using gcc libraries :: <toolset>gcc-5.4.1 :: C:/Qt/Tools/mingw810_64/bin C:/Qt/Tools/mingw810_64/lib C:/Qt/Tools/mingw810_64/lib32 C:/Qt/Tools/mingw810_64/lib64
notice: using gcc archiver :: <toolset>gcc-5.4.1 :: C:/Qt/Tools/mingw810_64/bin/ar.exe
notice: using gcc ranlib :: <toolset>gcc-5.4.1 :: C:/Qt/Tools/mingw810_64/bin/ranlib.exe
notice: using rc compiler :: <toolset>gcc-5.4.1 :: C:\Qt\Tools\mingw810_64\bin\windres.exe
...found 119 targets...
...updating 10 targets...
gcc.compile.c++ C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\boost\build\31539694badee8abff083ccb1877aba1\error_code.o
gcc.link.dll.mingw C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\boost\build\31539694badee8abff083ccb1877aba1\libboost_system.dll.a
common.copy C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\stage\lib\libboost_system.dll.a
C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\boost\build\31539694badee8abff083ccb1877aba1\libboost_system.dll.a
        1 file(s) copied.
common.copy C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\stage\lib\libboost_system.dll
C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel\boost\build\31539694badee8abff083ccb1877aba1\libboost_system.dll
        1 file(s) copied.
...updated 10 targets...
[1/2] cmd.exe /C "cd /D "C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel" && "C:\foo bar\vcpkg\downloads\tools\cmake-3.19.2-windows\cmake-3.19.2-win32-x86\bin\cmake.exe" -P cmake_install.cmake"
-- Install configuration: "Release"

</pre></details>

```
$ ls C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-dbg\stage\lib
libboost_system.dll
libboost_system.dll.a
```

```
$ ls C:\foo bar\vcpkg\buildtrees\boost-system\x64-mingw-dynamic-rel
vcpkg-parallel-configure
```

**Comment by StarGate-One** [comment]
@autoantwort, yes, looks like the --build-dir --stage-dir --user-config and other options being passed to b2 within the quotes instead of just the PATH being inside quotes i.e.
_"--user-config=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel/user-config.jam" should be_
*--user-config="C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel/user-config.jam"*
_"--stagedir=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel/stage" should be_
*--stagedir="C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel/stage"*
_"--build-dir=C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel" should be_
*--build-dir="C:/foo bar/vcpkg/buildtrees/boost-system/x64-mingw-dynamic-rel"*
_"-sBOOST_ROOT=C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build" should be_
*-sBOOST_ROOT="C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build"*
_"-sBOOST_BUILD_PATH=C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build" should be_
*-sBOOST_BUILD_PATH="C:/foo bar/vcpkg/installed/x64-mingw-dynamic/tools/boost-build"*

at least how I interpret the boost b2 documentation:
The command prompt treats each bit of whitespace in the command as an argument separator. That means quotation marks (") are required to keep text together whenever a single command-line argument contains spaces, as in
--build-dir="C:\Documents and Settings\dave\build-boost"

Then the whole command line would need to be triple quoted or escape characters used to properly pass this command with spaces within the PATHs.
This is one of the many reasons I do not use ^spaces^ in PATHs I can control - there is just too many tools/utilities that still do not support it. I wish the all wise ones would have never allowed a space to be used.

**Comment by longnguyen2004** [comment]
@StarGate-One Yes, a lot of build systems still can't handle spaces in paths properly.

**Comment by longnguyen2004** [comment]
Issue cleanup: Close as "unfixable"
